test/opt/test_eliminate_dead_const.cpp

```diff
@@ -0,0 +1,717 @@
+        // constants while other are not.
+        {
+            /* .used_consts = */
+            {
+                "%used_int = OpConstant %int 1",
+                "%used_double = OpConstant %double 3.14159265358979",
+            },
+            /*  .main_insts = */
+            {
+                "%int_var = OpVariable %_pf_int Function",
+                "OpStore %int_var %used_int",
+                "%double_var = OpVariable %_pf_double Function",
+                "OpStore %double_var %used_double",
+            },
+            /* .dead_consts = */
+            {
+                "%dead_bool = OpConstantTrue %bool",
+                "%dead_float = OpConstant %float 2.5",
+                "%dead_struct = OpConstantComposite %flat_struct %dead_bool %used_int %dead_float %used_double",
+            },
+        },
+        // A nesting struct type dead constants. All of its components are
+        // dead and should be removed after dead constant elimination.
+        {
+            /* .used_consts = */
+            {
+            },
+            /* .main_insts = */ {},
+            /* .dead_consts = */
+            {
+                "%dead_bool = OpConstantTrue %bool",
+                "%dead_int = OpConstant %int 1",
+                "%dead_float = OpConstant %float 2.5",
+                "%dead_double = OpConstant %double 3.1415926535",
+                "%dead_inner_struct = OpConstantComposite %inner_struct "
+                "%dead_bool %dead_int %dead_float %dead_double",
+                "%dead_int2 = OpConstant %int 2",
+                "%dead_double2 = OpConstant %double 1.428571428514",
+                "%dead_outer_struct = OpConstantComposite %outer_struct %dead_inner_struct %dead_int2 %dead_double2",
+            },
+        },
+        // A nesting struct type dead constants. Some of its components are
+        // dead constants while other are not.
+        {
+            /* .used_consts = */
+            {
+                "%used_int = OpConstant %int 1",
+                "%used_double = OpConstant %double 3.14159265358979",
+            },
+            /*  .main_insts = */
+            {
+                "%int_var = OpVariable %_pf_int Function",
+                "OpStore %int_var %used_int",
+                "%double_var = OpVariable %_pf_double Function",
+                "OpStore %double_var %used_double",
+            },
+            /* .dead_consts = */
+            {
+                "%dead_bool = OpConstantTrue %bool",
+                "%dead_float = OpConstant %float 2.5",
+                "%dead_inner_struct = OpConstantComposite %inner_struct %dead_bool %used_int %dead_float %used_double",
+                "%dead_int = OpConstant %int 2",
+                "%dead_double = OpConstant %double 1.428571428514",
+                "%dead_outer_struct = OpConstantComposite %outer_struct %dead_inner_struct %dead_int %dead_double",
+            },
+        },
+        // clang-format on
```

Comment:
It would be nice to see a test case in which both the inner and outer struct have dead and used elements.